### ISSUE TITLE
cross-compile with prerequisite tools

### ISSUE THREAD

**Issue by Johnnyxy** [conan-io/conan#5238]
Hi there,
what would be the best practise for the following use case?

I have a cross-compile scenario where I have to build a library (LIB_A) on Windows for QNX.
I have to run a tool (TOOL_A) while building LIB_A. I have the source for the TOOL_A and can compile it with Visual Studio.

TOOL_A depends on nothing and just needs to know where it has to run (os_build: Windows, arch_build: x86/x86_64).
But at build-time of LIB_A the compiler is QCC and Conan tries to build TOOL_A with QCC.

This means that Conan applies the setting "compiler" to the dependencies of LIB_A -> TOOL_A.
But TOOL_A has to be built with Visual Studio (on Windows).

What would be the best practise here?

At the moment I can just imagine that one has to ensure that the TOOL_A packages are available before building LIB_A.

Best regards

**Comment by danimtb** [comment]
Hi @Johnnyxy,

Yes, this is one of the flaws of our current crossbuilding model. When you face this kind of situation where the compiler changes you will probably have to use different profiles and do the crossbuilding in two steps: FIrst build the tool, then compile LIB_A with the pre-built tool.

I have added this issue to the cross-building project where you will find information about similar issues we are aware of.

We are currently working on providing a solution for the cross-building model based on the idea of applying the settings taking the context of the build into account (building a tool for a host machine, building a lib for a different target...).

Thanks for reporting.

**Comment by jgsogo** [comment]
Yes, this is something I'm targetting for the cross-building feature (WIP: https://github.com/conan-io/conan/pull/5202). For that PR you will need a profile for `host` package and one for `build` packages.

Current implementation follows these lines:

Profiles required
 * `profile_host` (the one you are already using)
    ```ini
    [settings]
    os=QNX
    compiler=qcc
    ... options, env
    ```

 * `profile_build` (the one for the build requires in the `build` context)
    ```ini
    [settings]
    os=Windows
    compiler=Visual Studio
    ... options, env
    ```

Then, your recipes should look like the following (we get rid of the `os_build` and `arch_build` settings):

```python
class ToolA(ConanFile):
    name = "tool_A"
    os = "os", "arch", "compiler", "build_type"

class LibA(ConanFile):
    name = "libA"
    os = "os", "arch", "compiler", "build_type"

    def build_requirements(self):
        self.build_requires("tool_A/xxx@user/channel", context="build")
```

and use both profiles from the command line:

```
conan create --profile_host=profile_host --profile_build=profile_build libA/conanfile.py user/channel
```

I'm starting to test it using `emsdk` and I have to say it is starting to work, but there is still some work ahead, we would like to release it asap but I cannot promise anything (end of June? 😉).

**Comment by Johnnyxy** [comment]
Hi, that is good, that such a feature is not postponed until Conan 2.0.

@jgsogo 
This is no criticism of the proposed feature.
Just a thought, can you imagine a scenario where one wants to supply a profile for a specific build_reuirement/requirement? Thus that would render the ```profile_host``` insufficient.

**Comment by jgsogo** [comment]
It could be an interesting feature but I'm not sure if it is related to the _cross-building feature_.

Right now Conan lets you specify particular settings for a reference using special syntax like `-s zlib:build_type=Release` that also works inside the profile files. For consistency it would require something like `--zlib:profile=filename` that the `argparse` wouldn't support (but I'm sure there could be some workaround like adding a fallback function to parse unknows arguments).

I think it is a feature that is worth considering on its own, but it probably has a lot of corner cases and can be really hard/complex to implement (although I agree that from the UX perspective it could be very helpful).

**Comment by Johnnyxy** [comment]
For the moment I can circumvent my problem.
As the TOOL I have to build is used at build time of the LIB.
This means the TOOL only depends on os_build/arch_build/compiler. And in the ```package_id``` function I remove the ```self.info.settings.compiler``` as this does not affect the final package id what has been used to built the TOOL.

TOOL -> os_build, arch_build
In my case I have to build TOOL on Windows with Cygwin and execute Visual Studio compiler in the Cygwin environment. This means I have to set the ```settings.arch_build``` as "arch" in 
```python
def build(self):
    ...
    with tools.vcvars(self.settings, arch=str(self.settings.arch_build)):
        ... execute commands to build ...
    ...
```
and set the ```target``` and ```host``` parameter of the AutoToolsBuildEnvironment helper's [configure](https://docs.conan.io/en/latest/reference/build_helpers/autotools.html?highlight=pthread#configure) function to ```False``` if I build on Linux.
```python
def build(self):
    ...
    autotools = AutoToolsBuildEnvironment(self)
    autotools.configure(configure_dir='..', host=False, target=False)
    autotools.make()
    autotools.install()
    ...
```

Essentially forcing the build helpers to use ```self.settings.arch_build``` instead of ```self.settings.arch```.

**Comment by jgsogo** [comment]
The proposed feature will be released in `1.24` 🎉  https://github.com/conan-io/conan/pull/5594
